### ISSUE TITLE
Prompt caching returns different logits for repeat prompts

### ISSUE THREAD

**Issue by cbowdon** [ml-explore/mlx-lm#259]
Hi, thanks for your work on this awesome project. I have been using the intelligent prompt caching described in #137 and implemented in https://github.com/ml-explore/mlx-lm/commit/60c93b8caa9923d9da7ec712bb3a9d62c521e7a2. Surprisingly, it does not return the same logits when repeating a prompt.

This is problematic for applications that use the logits, because you will then get different logits for the same request. In my experience it is not usually unstable enough to affect token sampling.

Here's a reproducible example adapted from the existing prompt cache tests. The first test trims the cache back to the first generated token and passes - more or less exactly the same as the existing test. The second trims back one token further so we provide the last token of the prompt, as is done [here in server.py](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/server.py#L571). However, the test fails because the logits change.

This only appears for longer prompts. The second test would pass if the prompt was `"this is a prompt"` like it is in the existing tests.

## Passing test

This is almost exactly the same as the prompt cache test [here](https://github.com/ml-explore/mlx-lm/blob/main/tests/test_prompt_cache.py#L210) except:

- the prompt is longer
- more tokens are generated to avoid chance agreement
- using `assert` for convenience with an interpreter rather than `self.assert*`

```python
HF_MODEL_PATH = "mlx-community/Qwen1.5-0.5B-Chat-4bit"

def test_trim_cache_with_generate(MODEL=HF_MODEL_PATH):
    model, tokenizer = load(MODEL)
    prompt = tokenizer.encode(
        "this is a prompt that's a bit longer", return_tensors="mlx"
    )[0]

    prompt_cache = make_prompt_cache(model)

    # Generate one token so we process the full prompt
    last_tok, _ = next(generate_step(prompt, model, prompt_cache=prompt_cache))
    last_tok = mx.array([last_tok])

    # Generate n more tokens
    n = 10  # use a higher n to avoid chance agreement of toks
    results = zip(range(n), generate_step(last_tok, model, prompt_cache=prompt_cache))
    toks, all_logits = zip(*(r[1] for r in results))

    # To get back to the cache just after processing the prompt,
    # trim by n + 1 tokens
    trim_prompt_cache(prompt_cache, n + 1)

    # Generate the same thing again
    results = zip(range(n), generate_step(last_tok, model, prompt_cache=prompt_cache))
    second_toks, second_all_logits = zip(*(r[1] for r in results))
    assert toks == second_toks
    assert all(mx.allclose(l, l2) for l, l2 in zip(all_logits, second_all_logits))
```

## Failing test

This is very close to the above, except:

- it generates `n` tokens directly
- it starts the cached generation from the last token of the prompt

This simulates the behaviour in `server.py`.

```python
def test_trim_cache_repeat_prompt(MODEL=HF_MODEL_PATH):
    model, tokenizer = load(MODEL)
    prompt = tokenizer.encode("this is a prompt that is longer", return_tensors="mlx")[0]

    prompt_cache = make_prompt_cache(model)

    # Generate n tokens
    n = 10  # use a higher n to avoid chance agreement of toks
    results = zip(range(n), generate_step(prompt, model, prompt_cache=prompt_cache))
    toks, all_logits = zip(*(r[1] for r in results))

    # To get back to the cache with the last token of the prompt
    # trim by n + 1 tokens
    trim_prompt_cache(prompt_cache, n + 1)

    # Generate the same thing again, n tokens after the end of the prompt
    results = zip(
        range(n),
        generate_step(prompt[-1:], model, prompt_cache=prompt_cache),
    )
    second_toks, second_all_logits = zip(*(r[1] for r in results))

    assert toks == second_toks, "Different tokens"

# THIS ASSERT FAILS
    assert all(
        mx.allclose(l, l2) for l, l2 in zip(all_logits, second_all_logits)
    ), "Different logits"
```

N.B. I have tried Qwen1.5, Qwen3, and Gemma3 with the same results. I've also tried for a range of generated tokens, different length prompts, and different amounts of shared prefix. As the prompt gets longer, the difference in logits becomes more obvious. 

I could reproduce the issue with mlx-lm versions 0.24.1 and 0.25.2.

Hoping you can help! Thanks in advance.

**Comment by awni** [comment]
This looks likely to be due to numerical differences between the two computations. When using low precision over several steps, these numerical differences can add up and eventually cause divergence between otherwise equivalent computations.

- First, I would rarely expect `mx.allclose` to pass with the default tolerance for `bfloat16` (or any half or lower precision). So I wouldn't over index on that.
- If you cast the model to float32 you should get the same tokens as output. I confirmed with `qwen/Qwen3-1.7B`.  Use `model.set_dtype(mx.float32)`
- In my case even at bfloat16 where the first output token differs the logits are still reasonably close. If you look at the selected location for the first and second pass they both get close scores but they switch in order. After that you would expect them to diverge rapidly.

I'm going to close this as the explanation makes sense. If you want to achieve numerically identical (or closer) results then your options are somewhat limited:
1. Don't use prompt caching
2. Use a higher precision

Although I would recommend avoiding the dependence on needing the results to be identical and make your downstream task more robust to that.

**Comment by cbowdon** [comment]
@awni thanks for the prompt response and suggestions! Would you be able to briefly explain why the numerical precision becomes a problem with prompt caching but not full computation please? Does prompt caching do some additional compression?

**Comment by awni** [comment]
I wouldn't describe it as a problem necessarily. It's just the sequence of operations are different. In the first case you process the full prompt in one go. In the second case you only process the last token (which you can think of as splitting the prompt into the first p-1 tokens followed by the last token and doing it in two steps).

**Comment by cbowdon** [comment]
Thank you, I think I understand... it's the cascading effect of lots of tiny numerical differences? Which would mean the divergence would be greater for longer prompts.

**Comment by awni** [comment]
> it's the cascading effect of lots of tiny numerical differences

Yea exactly.

> Which would mean the divergence would be greater for longer prompts.

Potentially yes. For example if you prefill the prompt token-by-token instead of all in one shot, you could end up with very different results just because of numerical differences between the two implementations compounding.

**Comment by cbowdon** [comment]
Makes sense, thank you!
